RFC/src/RFC-0201_TariScript.md

```diff
@@ -286,54 +487,247 @@ The proposal for Tari Script is straightforward. It is based on Bitcoin script a
+
+Fails with `Failure` if:
+* there is not a valid integer value on top of the stack,
+* the stack is empty.
+
+### Stack manipulation
+
+##### PushZero
+
+Pushes a zero onto the stack. This is a very common opcode and has the same effect as `PushInt(0)` but is more compact.
+
+The default script is a single `PushZero`.
 
-### Serialisation
+##### PushOne
+
+Pushes a one onto the stack. This is a very common opcode and has the same effect as `PushInt(1)` but is more compact.
+
+`PushOne` can be used in conditionals and to represent `false` or a failure condition. For example,
+
+    PushOne
+
+is a burn script, meaning that no-one can spend the output, since the script will always fail, similar to `Return`
+
+##### PushHash(HashValue)
+
+Push the associated 32-byte value onto the stack.
+
+Fails with `Failure` if:
+* HashValue is not a valid 32 byte sequence
+* the stack would exceed the max stack height.
+
+##### PushInt(i64)
+
+Push the associated 64 bit signed integer onto the stack
+
+Fails with `Failure` if:
+* HashValue is not a valid 8 byte sequence
+* the stack would exceed the max stack height.
+
+##### PushPubKey(PublicKey)
+
+Push the associated 32-byte value onto the stack. It will be interpreted as a public key or a commitment.
+
+Fails with `Failure` if:
+* PublicKey is not a valid 32 byte sequence
+* the stack would exceed the max stack height.
+
+##### Drop
+
+Drops the top stack item.
+
+Fails with `Failure` if:
+* The stack is empty
+
+##### Dup
+
+Duplicates the top stack item.
+
+Fails with `Failure` if:
+* The stack is empty
+* the stack would exceed the max stack height.
+
+##### RevRot,
+
+Reverse rotation. The top stack item moves into 3rd place, e.g. `abc => bca`.
+
+Fails with `Failure` if:
+* The stack has two or fewer elements
+
+### Math operations
+
+##### Add
+
+Pop two items and push their sum
+
+Fails with `Failure` if:
+* The stack has a height of one or less
+
+##### Sub
+
+Pop two items and push the second minus the top
+
+Fails with `Failure` if:
+* The stack has a height of one or less
+
+##### Equal
+
+Pops the top two items, and pushes 0 to the stack if the inputs are exactly equal, 1 otherwise.
+
+Fails with `Failure` if:
+* The stack height is one or less
+
+##### EqualVerify
+
+Pops the top two items, and compare their values.
+
+Fails with `Failure` if:
+* The stack height is one or less,
+* The top two stack elements are not equal
+
+### Boolean logic
+
+#### Or(n)
+
+`n` items are popped from the stack. The top item must match at least one of those items.
+
+Fails with `Failure` if:
+* The stack has height `n` or less
+* The top value does not match any of the popped items
+
+### Cryptographic operations
+
+##### HashBlake256
+
+Pop the top element, hash it with the Blake256 hash function and push the result to the stack.
+
+Fails with `Failure` if:
+* The stack is empty
+
+##### CheckSig,
+
+Pop the public key and then the signature. If the signature signs the script, push 0 to the stack, otherwise
```

Comment:
Did not find it anywhere in the document, but how would signature get into the stack in first place? Initially thought that supposed to be part of input data, but then thought it's recursion that first you need to have input data to be able to create signature..

Reply:
In this case the signature is in the input data. Input data is not part of the script itself, so it's possible to sign it without recursion issues.

But I do acknowledge that the definition of what's being signed is not well defined here and should be addressed.